### PR TITLE
Adds ability to customize DomainMessage getType method

### PR DESCRIPTION
**Pull request by aleixverges**
- Creates an EventStream decorator to allow DomainMessage type customization
- Updates DomainMessage to allow type update

### CODE REVIEW

File: src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    public function decorateForWrite($aggregateType, $aggregateIdentifier, DomainEventStreamInterface $eventStream)
+    {
+        $messages = array();
+
+        foreach ($eventStream as $message) {
+            $payload = $message->getPayload();
+            if (method_exists($payload, 'getType')) {
```
Comment (unkind) [inline src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php]: What if event accidentally contains this getter?

File: src/Broadway/Domain/DomainMessage.php [hunk 2]
```diff
@@ -131,4 +140,15 @@ public function andMetadata(Metadata $metadata)
...
         return new DomainMessage($this->id, $this->playhead, $newMetadata, $this->payload, $this->recordedOn);
     }
+
+    /**
+     * @param string $type
+     * @return $this
+     */
+    public function andType($type)
+    {
+        $this->type = $type;
```
Comment (unkind) [inline src/Broadway/Domain/DomainMessage.php]: This makes `DomainMessage` mutable.

File: src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php (hunk 1 above)
Comment (aleixverges) [inline src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php]: That's more less the plan. A new EventInterface could exist with the getType or messageName method defined. The decorator would use that to create new DomainMessage with that new message key.

File: src/Broadway/Domain/DomainMessage.php (hunk 2 above)
Comment (aleixverges) [inline src/Broadway/Domain/DomainMessage.php]: A new optional parameter could be added to DomainMessage class called "type" for example. That would make the creation of a new DomainMessage, like with the andMetadata decorator, possible

**Comment by aleixverges** [comment]
Thanks for the feedback!
      I can prepare an updated pull request to show the changes in the implementation fixing the issues you saw.

Aleix

File: src/Broadway/Domain/DomainMessage.php (hunk 2 above)
Comment (unkind) [inline src/Broadway/Domain/DomainMessage.php]: You don't have to use constructor:
```
public function andType($type)
{
    $message = clone $this;
    $message->type = $type;

    return $message;
}
```

**Comment by aleixverges** [comment]
Hi,
    I've updated the pull request with some changes according your comments.
Thanks!
Aleix

File: src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+     * {@inheritDoc}
+     */
+    public function decorateForWrite($aggregateType, $aggregateIdentifier, DomainEventStreamInterface $eventStream)
+    {
+        $messages = array();
+
+        /** @var DomainMessage $message */
+        foreach ($eventStream as $message) {
+            $payload = $message->getPayload();
+            if (method_exists($payload, 'getType')) {
```
Comment (othillo) [inline src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php]: if ($payload implements EventInterface) ?

**Pull request closed without merging**